Create CSPP page

pLabarta opened the pull request:
Closes #1031

RichardRed0x commented on docs/privacy/cspp.md:
```diff
@@ -0,0 +1,79 @@
+    - mixchange=1 - This tells dcrwallet to mix payments to changeaccount into the mixedaccount.
+    - ticketbuyer.votingaccount=voting - This tells the wallet which account to use when setting voting ticket voting addresses.
+
+In addition to setting these new options in `dcrwallet.conf`, **two new accounts must be created: mixed and unmixed.** The voting account can either be created locally, if your `ticketbuyer` wallet also votes, or you can import an extended pubkey for a voting account from an external voting wallet.
```
Could use some more detail on how to export and import xpub

pLabarta replied:
I'll add the cspp.decred.org and detailed info about exporting and importing xpubs in the two sections mentioned above.

dajohi replied:
The initial reasoning for this page was so the cspp daemon didn't need to run http.

RichardRed0x replied:
RIght, so this page should reproduce all of the useful practical info on cspp.decred.org then.

pLabarta commented on docs/privacy/cspp.md:
```diff
@@ -0,0 +1,79 @@
+
+In addition to setting these new options in `dcrwallet.conf`, **two new accounts must be created: mixed and unmixed.** The voting account can either be created locally, if your `ticketbuyer` wallet also votes, or you can import an extended pubkey for a voting account from an external voting wallet.
+
+Stakeholders running a ticketbuyer will want to setup a **2nd ticketbuyer** that will run in **parallel**, where the configurations are such that newly bought tickets in the 1st wallet have the 2nd wallet’s mixed account branch 0 set as their mixed account.
```
Yes, it is for those wanting to transition from an unmixed ticket buyer. To make it clear, I'll split the HowTo section in two: a) transitioning from unmixed to mixed and b) how non-stakers can mix received funds.

jholdstock replied:
I think these could be split into different pages

- What is CSPP, motivation, how it works, its benefits/drawbacks
- How-to guides - how to use cspp, how to transition to a private ticket buyer, etc

pLabarta replied:
Should it be a dropdown like PoS Voting / PoW Mining?

- CoinShuffle++
    + Overview
    + Using the mixnet

jholdstock replied:
Yeah I think that is sensible. Although I would probably use the following names:

- CoinShuffle++ (CSPP)
    - Overview
    - How to use CSPP

jholdstock commented on docs/privacy/cspp.md:
```diff
@@ -0,0 +1,79 @@
+---
+
+## How do I use CSPP?
+
```
The info on cspp.decred.org should eventually be copied into the docs, so we should not link out to that URL

pLabarta commented:
Splitted pages, added motivation section and the the LockEye.svg from the latest icon release to the dcr-icons folder.

jholdstock commented on docs/privacy/cspp/how-to-cspp.md:
````diff
@@ -0,0 +1,103 @@
+  --mixedaccount=mixed/1 --changeaccount=unmixed --mixchange
+```
+
+Alternatively, the mixaccount JSON-RPC may be used instead of the --mixchange option to mix single outputs from the account without leaving the wallet persistently unlocked.
````
Wrap `--mixchange` flag in backticks please

jholdstock commented on docs/privacy/cspp/overview.md:
```diff
@@ -0,0 +1,73 @@
+
+---
+
+## Limitations
```
We can add the following to Limitations (taken from cspp.decred.org)

> Address reusage strips the anonymity provided by CoinShuffle++. It is imperative that addresses are never reused and that extended public keys of mixed and voting accounts are not revealed to other parties.

jholdstock commented on docs/privacy/cspp/overview.md:
```diff
@@ -0,0 +1,73 @@
+
+Beyond the concern about pruning, other implementations carry moderate or high complexity, making code audits costly and broadening the attack surface.
+
+TumbleBit, from  [“TumbleBit: An Untrusted Bitcoin-Compatible Anonymous Payment Hub”](https://eprint.iacr.org/2016/575.pdf), was the first option. But the development team noticed that the mixing server would be vulnerable to Denial-of-Service ('DoS') attacks, with the countermeasures adding more complexity. Despite not integrating the TumbleBit code into Decred, the code has been [released](http ...
```
PR has been merged. This URL can be updated to https://decred.org/research/heilman2016.pdf

pLabarta commented:
Modified both pages according to @jholdstock requested changes. I replaced the TumbleBit URL with decred.org/research one but it does not seem to be live yet, it returns a 404.

I paraphrased Limitations, Benefits and the Centralized Server section. Hope it reads better. And added a Handling malicious peers subsection (there might be a better title for it).

jholdstock commented on docs/privacy/cspp/overview.md:
```diff
@@ -0,0 +1,85 @@
+
+Other privacy implementations like zk-SNARKS and Ring Signatures provide substantial privacy, but impede the network from dropping historical transactions from their full nodes, a process also known as pruning. The CSPP mixnet does not have this effect on the blockchain and pruning remains possible. Pruning the blockchain reduces its size, which makes it easier to download and replicate.
+
+At the same time, the CSPP mixnet presents a much simpler solution that is based on common cryptographic primitives and finite field arithmetic. It is simpler and more DoS resistant than [TumbleBit](https://eprint.iacr.org/2016/575.pdf), which was initially planned and developed,but dropped because of its vulnerability to Denial-of-Service attacks.
```
One more place to update the eprint URL to the decred.org one.

jholdstock commented on docs/privacy/cspp/how-to-cspp.md:
````diff
@@ -0,0 +1,103 @@
+  --mixedaccount=mixed/1 --changeaccount=unmixed --mixchange
+```
+
+Alternatively, the mixaccount JSON-RPC may be used instead of the `--mixchange` option to mix single outputs from the account without leaving the wallet persistently unlocked.
````
Sorry I missed this the first time - The name of the RPC `mixaccount` should also be in backticks

jholdstock commented on docs/privacy/cspp/overview.md:
```diff
@@ -36,35 +36,47 @@ The team ended up pivoting to CoinShuffle++, a simpler and more DoS-resistant so
+
+### Handling malicious peers
+
+The protocol handles collisions and malicious peers using a privacy-preserving redundant messaging step. In its worst case scenario, it requires `4+2f` communication rounds for `f` malicious peers. To achieve this, it that each peer uses a freshly generated bitstring (verification key).
```
> To achieve this, it that each peer uses a freshly generated bitstring (verification key).

Bad grammar here, I'm guessing typo. Missing word?

jholdstock commented on docs/privacy/cspp/overview.md:
```diff
@@ -36,35 +36,47 @@ The team ended up pivoting to CoinShuffle++, a simpler and more DoS-resistant so
 ## Benefits
 
-The implementation of the CSPP mixnet did not require changing the consensus rules. It follows Decred's low-hanging fruit approach and it allowed Decred developers to work on it in a permisionless manner.
+The implementation of the CSPP mixnet did not require changing the consensus rules. It is an opt-in implementation that  obfuscates ownership of DCR coins with out requiring any modifications to the consensus rules. A non-opt-in system would be much more complex.
```
> A non-opt-in system would be much more complex."

I think this extra sentence should be removed unless some extra justification is added. Discussing the pro and cons of opt-in/opt-out is a whole new can of worms.

jholdstock commented on docs/privacy/cspp/overview.md:
```diff
@@ -36,35 +36,47 @@ The team ended up pivoting to CoinShuffle++, a simpler and more DoS-resistant so
-The initial code only supports the CLI wallet, dcrwallet, and solo stakers. It does not work with voting service providers (VSPs), and regular transactions.
+The current implementation has several limitations. Mainly the use of a centralized server, to which the CoinJoin leaks which inputs and change addresses belong to each peer, and the lack of support for Voting Service Providers and regular transactions.
+
+Change and address use also requires careful handling. Address reusage strips the anonymity provided by CoinShuffle++. It is imperative that addresses are never reused and that extended public keys of mixed and voting accounts are not revealed to other parties.
```
> Change and address use also requires careful handling. Address reusage strips the anonymity provided by CoinShuffle++.

Suggested rewording:

> The anonymity provided by CoinShuffle++ can be lost if wallet addresses are reused, or if change outputs are not handled correctly.

pLabarta commented:
There! Fixed those! Thank you @jholdstock !

jholdstock commented:
Looks good to merge once the conflict is resolved.
